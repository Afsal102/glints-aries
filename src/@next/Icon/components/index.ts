export { default as RiAccountCircleFill } from './RiAccountCircleFill';
export { default as RiAccountCircleLine } from './RiAccountCircleLine';
export { default as RiAddCircleFill } from './RiAddCircleFill';
export { default as RiAddCircleLine } from './RiAddCircleLine';
export { default as RiAdd } from './RiAdd';
export { default as RiAlarmWarningFill } from './RiAlarmWarningFill';
export { default as RiAlarmWarningLine } from './RiAlarmWarningLine';
export { default as RiArrowDownFill } from './RiArrowDownFill';
export { default as RiArrowDownLine } from './RiArrowDownLine';
export { default as RiArrowGoBackFill } from './RiArrowGoBackFill';
export { default as RiArrowGoBackLine } from './RiArrowGoBackLine';
export { default as RiArrowGoForwardFill } from './RiArrowGoForwardFill';
export { default as RiArrowGoForwardLine } from './RiArrowGoForwardLine';
export { default as RiArrowLeftFill } from './RiArrowLeftFill';
export { default as RiArrowLeftLine } from './RiArrowLeftLine';
export { default as RiArrowLeftRightFill } from './RiArrowLeftRightFill';
export { default as RiArrowLeftRightLine } from './RiArrowLeftRightLine';
export { default as RiArrowMDownFill } from './RiArrowMDownFill';
export { default as RiArrowMDownLine } from './RiArrowMDownLine';
export { default as RiArrowMLeftFill } from './RiArrowMLeftFill';
export { default as RiArrowMLeftLine } from './RiArrowMLeftLine';
export { default as RiArrowMRightFill } from './RiArrowMRightFill';
export { default as RiArrowMRightLine } from './RiArrowMRightLine';
export { default as RiArrowMUpFill } from './RiArrowMUpFill';
export { default as RiArrowMUpLine } from './RiArrowMUpLine';
export { default as RiArrowRightFill } from './RiArrowRightFill';
export { default as RiArrowRightLine } from './RiArrowRightLine';
export { default as RiArrowUpDownFill } from './RiArrowUpDownFill';
export { default as RiArrowUpDownLine } from './RiArrowUpDownLine';
export { default as RiArrowUpFill } from './RiArrowUpFill';
export { default as RiArrowUpLine } from './RiArrowUpLine';
export { default as RiArrowXsDownFill } from './RiArrowXsDownFill';
export { default as RiArrowXsUpFill } from './RiArrowXsUpFill';
export { default as RiCalendarFill } from './RiCalendarFill';
export { default as RiCalendarLine } from './RiCalendarLine';
export { default as RiCheck } from './RiCheck';
export { default as RiCloseCircleFill } from './RiCloseCircleFill';
export { default as RiCloseCircleLine } from './RiCloseCircleLine';
export { default as RiClose } from './RiClose';
export { default as RiDeleteBinFill } from './RiDeleteBinFill';
export { default as RiDeleteBinLine } from './RiDeleteBinLine';
export { default as RiErrorWarningFill } from './RiErrorWarningFill';
export { default as RiErrorWarningLine } from './RiErrorWarningLine';
export { default as RiInformationFill } from './RiInformationFill';
export { default as RiInformationLine } from './RiInformationLine';
export { default as RiMessageFill } from './RiMessageFill';
export { default as RiMessageLine } from './RiMessageLine';
export { default as RiPencilFill } from './RiPencilFill';
export { default as RiPencilLine } from './RiPencilLine';
export { default as RiQuestionFill } from './RiQuestionFill';
export { default as RiQuestionLine } from './RiQuestionLine';
export { default as RiReplyFill } from './RiReplyFill';
export { default as RiReplyLine } from './RiReplyLine';
export { default as RiSendPlaneFill } from './RiSendPlaneFill';
export { default as RiSendPlaneLine } from './RiSendPlaneLine';
export { default as RiSettingsFill } from './RiSettingsFill';
export { default as RiSettingsLine } from './RiSettingsLine';
