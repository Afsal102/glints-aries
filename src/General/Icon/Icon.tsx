import * as React from 'react';

const Icon: React.FunctionComponent<Props> = props => {
  const {
    className,
    children,
    color,
    width = '1em',
    height = '1em',
    onClick,
    ...restProps
  } = props;

  return (
    <svg
      className={className}
      width={width}
      height={height}
      onClick={onClick}
      fill={color}
      viewBox="0 0 100 100"
      {...restProps}
    >
      {children}
    </svg>
  );
};

export interface Props {
  className?: string;
  children: React.ReactNode;
  color?: string;
  width?: string | number;
  height?: string | number;
  onClick?(e: React.MouseEvent<SVGSVGElement, MouseEvent>): void;
}

export default Icon;
