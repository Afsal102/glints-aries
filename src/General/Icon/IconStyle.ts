import styled from 'styled-components';

export const VerticalCenteredSvg = styled.svg`
  vertical-align: middle;
  ${({ fontSize }) => {
    switch (fontSize) {
      case 'xxs':
        return `
          font-size: 12px;
        `;
      case 'xs':
        return `
          font-size: 16px;
        `;
      case 's':
        return `
          font-size: 24px;
        `;
      case 'm':
        return `
          font-size: 32px;
        `;
      case 'l':
        return `
        font-size: 48px;
        `;
      case 'xl':
        return `
        font-size: 64px;
        `;
      case 'xxl':
        return `
        font-size: 96px;
        `;
      case 'xxxl':
        return `
        font-size: 128px;
        `;
      default:
        return `
        font-size: 24px;
        `;
    }
  }}
`;
