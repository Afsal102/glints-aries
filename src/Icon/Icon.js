// @flow
import React from 'react';
import styled, {css} from 'styled-components';
import View from '../View';
import type {Width} from '../View/View';
import StyleMatcher from '../Utils/Matcher';

type IconSize = 'default' | 'medium' | 'small' | 'xsmall';
type IconBgColor =
  | 'regular'
  | 'cyan'
  | 'gray'
  | 'red'
  | 'green'
  | 'blue'
  | 'orange'
  | 'white'
  | 'purple';

type Props = {
  size?: IconSize,
  children: React$Element<'img' | 'svg' | 'i'>,
  padding?: Width,
  paddingLeft?: Width,
  paddingRight?: Width,
  backgroundColor?: IconBgColor,
  rounded?: boolean,
};

const iconSizeHandler = StyleMatcher.create('size', [
  ['default', '2.4rem'],
  ['medium', '2.0rem'],
  ['small', '1.8rem'],
  ['xsmall', '1.2rem'],
]);

const IconWrapper = styled.span`
  display: block;
  position: relative;
  ${props =>
    props.backgroundColor &&
    css`
      background-color: ${StyleMatcher.create('backgroundColor', [
        ['regular', '#000000'],
        ['cyan', '#00a1b0'],
        ['gray', '#999999'],
        ['red', '#f44336'],
        ['green', '#4CAF50'],
        ['blue', '#2196F3'],
        ['orange', '#FF9800'],
        ['white', '#FFFFFF'],
        ['purple', '#673AB7'],
      ])};
    `};
  ${props =>
    props.rounded &&
    css`
      border-radius: 50%;
    `};
`;

const IconContentWrapper = styled.div`
  width: ${iconSizeHandler};
  height: ${iconSizeHandler};
`;

const Icon = ({children, padding, paddingRight, paddingTop, ...iconProps}: Props) => (
  <IconWrapper {...iconProps}>
    <View padding={padding} paddingRight={paddingRight} paddingTop={paddingTop}>
      <IconContentWrapper {...iconProps}>
        {React.cloneElement(children, {
          style: {
            display: 'block',
            width: iconSizeHandler(iconProps),
            height: iconSizeHandler(iconProps),
          },
        })}
      </IconContentWrapper>
    </View>
  </IconWrapper>
);

Icon.defaultProps = {
  size: 'default',
};

export type {IconSize, IconBgColor};
export default Icon;
