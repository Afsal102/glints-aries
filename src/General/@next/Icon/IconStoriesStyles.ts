import styled from 'styled-components';

import { Icon, IconProps } from './Icon';

export const StyledUnorderedList = styled.ul`
  display: flex;
  list-style-type: none;
  flex-wrap: wrap;
  align-items: flex-start;
`;

export const StyledItemList = styled.li`
  display: flex;
  flex-basis: 220px;
  align-items: center;
`;

export const StyledIcon = styled(Icon)<IconProps>`
  margin-right: 1em;
`;
