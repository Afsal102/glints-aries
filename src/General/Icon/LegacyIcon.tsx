import * as React from 'react';
import camelCase from 'lodash/camelCase';
import upperFirst from 'lodash/upperFirst';

import { ICONS } from '../../Utils/Icon/IconLibrary';

const formatWarning = (name: string) => {
  const camelCaseName = upperFirst(camelCase(name));

  return `Warning: <Icon name="${name}" /> will be deprecated in a future release.\n\nPlease use <${camelCaseName}Icon /> to instead. See http://aries.glints.design/?path=/story/general--icon for details.`;
};

const LegacyIcon: React.FunctionComponent<Props> = props => {
  const { name, color } = props;
  const iconName = name.replace(/-/g, '_');

  if (typeof console !== 'undefined') {
    console.warn(formatWarning(name));
  }

  return (
    <svg width="1em" height="1em" fill={color} viewBox="0 0 100 100">
      {(ICONS as any)[iconName.toUpperCase()]}
    </svg>
  );
};

export interface Props {
  name: string;
  color?: string;
}

export default LegacyIcon;
