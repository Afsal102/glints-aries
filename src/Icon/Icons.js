// @flow
import React from 'react';
import PropTypes from 'prop-types';
import type {Width} from '../View/View';
import Icon, {type IconSize, type IconBgColor} from './Icon';
import {UikitConsumer} from '../Provider';

type IconName =
  | 'add'
  | 'bold'
  | '18-plus'
  | 'bullet-list'
  | 'chevron-back'
  | 'chevron-next'
  | 'close'
  | 'close-white'
  | 'code'
  | 'comment-active'
  | 'comment'
  | 'comment-white'
  | 'create-story'
  | 'drop-down'
  | 'elipsis-menu'
  | 'embed'
  | 'image'
  | 'italic'
  | 'like-active'
  | 'like-active-white'
  | 'like'
  | 'like-white'
  | 'link'
  | 'logout'
  | 'notification'
  | 'number-list'
  | 'pin-map'
  | 'profile'
  | 'search'
  | 'setting'
  | 'share-to-timeline'
  | 'share'
  | 'signal'
  | 'share-white'
  | 'sosmed-facebook'
  | 'sosmed-instagram'
  | 'sosmed-linkedin'
  | 'sosmed-twitter'
  | 'sosmed-youtube'
  | 'underline'
  | 'verified'
  | 'verified-green';

type Props = {
  name: IconName,
  size?: IconSize,
  padding?: Width,
  backgroundColor?: IconBgColor,
  rounded?: boolean,
};

const Icons = ({name, ...iconProps}: Props) => (
  <UikitConsumer>
    {uikit => (
      <Icon {...iconProps}>
        <img src={require(`../../assets/icons/${name}.svg`)} alt={name} />
      </Icon>
    )}
  </UikitConsumer>
);

Icons.contextTypes = {
  uikitAssets: PropTypes.object,
};

export type {IconName};
export default Icons;
