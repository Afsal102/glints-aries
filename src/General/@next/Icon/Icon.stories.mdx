import { Meta } from '@storybook/addon-docs';
import { iconNames } from './icons/icons';
import { Typography } from '../Typography'
import { StyledIcon, StyledItemList, StyledUnorderedList } from './IconStoriesStyles'
import { Icon } from './Icon';
import * as Colors from '../../../Utils/@next/colors';

<Meta title="General/@next/Icon" />

# Icons

<StyledUnorderedList>
  {iconNames.map((name) =>
    <StyledItemList key={name}>
      <StyledIcon name={name} height={28}/>
      <Typography variant='subtitle2'>{name}</Typography>
    </StyledItemList>
  )}
</StyledUnorderedList>

## How to use
- import NIcon component from glints-aries and pass name prop to use the specific icon

<div style={{ display: 'flex' }}>
  <StyledIcon name="ri-account-circle-fill" height={48}/>
  <Typography variant='subtitle2'>ri-account-circle-fill</Typography>
</div>

```ts
import { NIcon as Icon } from 'glints-aries';

export const MyComponent = () => (
  <Icon name="ri-account-circle-fill" />
);
```

### Color
- you can change the color of the icon by passing the color to `fill` prop.

<StyledIcon name="ri-account-circle-fill" height={48} fill={Colors.Blue.Brand} />

```
Import { Colors } from 'glints-aries'

<Icon name="ri-account-circle-fill" height={48} fill={Colors.Blue.Brand} />
```

### Size
- you can change the size of the icon by either passing a height or width as **props** or using **styles**.
- passing at least either the `height` or `width` automatically sets the other with the same value.
#### Props
- Passing height or width as props
```ts
<Icon name="ri-account-circle-fill" height={48} />
```

#### Style
- Resizing by specifying height or width in styles
```ts
import { NIcon as Icon } from 'glints-aries';
import styled from 'styled-components';

const StyledIcon = styled(Icon)`
  height: '48px';
`;

export const MyComponent = () => (
  <StyledIcon name="ri-account-circle-fill" />
);
```

## Adding new icons
1. Add svg file to `src/General/@next/Icon/icons`
  - make sure only viewBox is defined. Remove height, width or fill. This is to make sure the generated React component generates correct props for styles to work. 
2. Generate SVG React Component. Run `yarn build`. Components can be found in `src/General/@next/Icon/components`
3. Update `src/General/@next/Icon/icons/icons.ts` file.
  - insert the new icon's name to `iconNames` in alphabetical order. Refer to Iconography naming convention.
  - update `iconsMappingComponent` with it's corresponding component name.
