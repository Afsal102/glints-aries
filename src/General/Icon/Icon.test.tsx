import * as React from 'react';

import * as renderer from 'react-test-renderer';
import { fireEvent, render } from '@testing-library/react';
import '@testing-library/jest-dom/extend-expect';

import Icon from './Icon';

const testId = 'icon-svg';
const children = (
  <path d="M100 56.563H56.562V100H43.438V56.562H0V43.438h43.438V0h13.124v43.438H100z"></path>
);

it('<Icon> should render a svg with color, width, height and a path as children', () => {
  const IconSnapshot = renderer.create(<Icon>{children}</Icon>).toJSON();
  expect(IconSnapshot).toMatchSnapshot();
});

it('<Icon> should accept a className to apply the style from styled-component', () => {
  const className = 'styled-class';
  const { getByTestId } = render(<Icon className={className}>{children}</Icon>);
  const icon = getByTestId(testId);
  expect(icon).toHaveClass(className);
});

it('<Icon> should be able to inherit color', () => {
  const { getByTestId } = render(<Icon>{children}</Icon>);
  const icon = getByTestId(testId);
  expect(icon).toHaveAttribute('fill', 'currentColor');
});

it('<Icon> should display the correct color prop', () => {
  const color = 'red';
  const { getByTestId } = render(<Icon color={color}>{children}</Icon>);
  const icon = getByTestId(testId);
  expect(icon).toHaveAttribute('fill', color);
});

it('<Icon> should call onClick once when it is clicked', () => {
  const onClick = jest.fn();
  const { getByTestId } = render(<Icon onClick={onClick}>{children}</Icon>);
  const icon = getByTestId(testId);
  fireEvent.click(icon);
  expect(onClick).toHaveBeenCalledTimes(1);
});
