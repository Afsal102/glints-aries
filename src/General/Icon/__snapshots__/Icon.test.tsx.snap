// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Icon> should matches snapshot 1`] = `
<svg
  className="IconStyle__VerticalCenteredSvg-sc-19ci3uf-0 dmBUgt"
  data-testid="icon-svg"
  fill="currentColor"
  height="1em"
  viewBox="0 0 100 100"
  width="1em"
>
  <path
    d="M100 56.563H56.562V100H43.438V56.562H0V43.438h43.438V0h13.124v43.438H100z"
  />
</svg>
`;
